File: obsidian-mcp-py/obsidian/embeds.py
"""
Obsidian Embeds and Blocks Management

Provides tools for managing note embeds and block-level operations.
"""

import re
from pathlib import Path
from typing import Optional, List, Dict, Any
from obsidian.vault import get_vault_path, safe_path


def extract_blocks(content: str) -> List[Dict[str, Any]]:
    """
    Extract block structure from note content.
    
    Blocks are identified by:
    - Headings (# ## ### etc.)
    - Code blocks (```)
    - Lists
    - Blockquotes (>)
    
    Args:
        content: Note content
        
    Returns:
        List of block dictionaries with type and content
    """
    blocks = []
    lines = content.splitlines()
    
    current_block = None
    block_id = 0
    
    for i, line in enumerate(lines):
        # Heading block
        if re.match(r'^#{1,6}\s', line):
            if current_block:
                blocks.append(current_block)
            
            level = len(line) - len(line.lstrip('#'))
            current_block = {
                "id": f"block_{block_id}",
                "type": "heading",
                "level": level,
                "content": line,
                "line_number": i + 1
            }
            block_id += 1
            
        # Code block start/end
        elif line.strip().startswith('```'):
            if current_block and current_block.get("type") == "code":
                current_block["content"] += "\n" + line
                blocks.append(current_block)
                current_block = None
            else:
                if current_block:
                    blocks.append(current_block)
                language = line.strip()[3:] if len(line.strip()) > 3 else ""
                current_block = {
                    "id": f"block_{block_id}",
                    "type": "code",
                    "language": language,
                    "content": line,
                    "line_number": i + 1
                }
                block_id += 1
        
        # Blockquote
        elif line.strip().startswith('>'):
            if current_block and current_block.get("type") == "blockquote":
                current_block["content"] += "\n" + line
            else:
                if current_block:
                    blocks.append(current_block)
                current_block = {
                    "id": f"block_{block_id}",
                    "type": "blockquote",
                    "content": line,
                    "line_number": i + 1
                }
                block_id += 1
        
        # Regular content
        else:
            if current_block:
                if current_block.get("type") in ["code", "blockquote"]:
                    current_block["content"] += "\n" + line
                else:
                    current_block["content"] += "\n" + line
            else:
                # Start a new paragraph block
                current_block = {
                    "id": f"block_{block_id}",
                    "type": "paragraph",
                    "content": line,
                    "line_number": i + 1
                }
                block_id += 1
    
    if current_block:
        blocks.append(current_block)
    
    return blocks


def extract_embeds(content: str) -> List[Dict[str, str]]:
    """
    Extract embedded notes from content.
    
    Embeds are in format: ![[Note Name]] or ![[Note Name#Heading]]
    
    Args:
        content: Note content
        
    Returns:
        List of embedded note dictionaries
    """
    embeds = []
    
    # Match ![[Note Name]] or ![[Note Name#Heading]]
    embed_pattern = r'!\[\[([^\]#]+)(?:#([^\]]+))?\]\]'
    matches = re.findall(embed_pattern, content)
    
    for match in matches:
        note_name = match[0].strip()
        heading = match[1].strip() if match[1] else None
        
        embeds.append({
            "note": note_name,
            "heading": heading,
            "embed_string": f"![[{note_name}{'#' + heading if heading else ''}]]"
        })
    
    return embeds


def register_embed_tools(mcp):
    """
    Register all embed and block-related MCP tools.
    
    Args:
        mcp: FastMCP instance to register tools with
    """
    
    @mcp.tool()
    def get_note_blocks(note_name: str) -> dict:
        """
        Extract block structure from a note.
        
        Args:
            note_name: Name of the note (without .md extension) or relative path
        
        Returns:
            Dictionary with block structure
        """
        try:
            vault_path = get_vault_path()
            note_path = safe_path(vault_path, f"{note_name}.md")
            
            if not note_path.exists():
                note_path = safe_path(vault_path, note_name)
            
            if not note_path.exists():
                return {"error": f"Note not found: {note_name}"}
            
            content = note_path.read_text(encoding="utf-8")
            blocks = extract_blocks(content)
            
            # Count blocks by type
            block_counts = {}
            for block in blocks:
                block_type = block["type"]
                block_counts[block_type] = block_counts.get(block_type, 0) + 1
            
            return {
                "note_name": note_path.stem,
                "total_blocks": len(blocks),
                "block_counts": block_counts,
                "blocks": blocks
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def update_block(note_name: str, block_id: str, new_content: str) -> dict:
        """
        Update a specific block in a note.
        
        Args:
            note_name: Name of the note (without .md extension) or relative path
            block_id: ID of the block to update (from get_note_blocks)
            new_content: New content for the block
        
        Returns:
            Dictionary with update status
        """
        try:
            vault_path = get_vault_path()
            note_path = safe_path(vault_path, f"{note_name}.md")
            
            if not note_path.exists():
                note_path = safe_path(vault_path, note_name)
            
            if not note_path.exists():
                return {"error": f"Note not found: {note_name}"}
            
            content = note_path.read_text(encoding="utf-8")
            blocks = extract_blocks(content)
            
            # Find the block to update
            block_to_update = None
            for block in blocks:
                if block["id"] == block_id:
                    block_to_update = block
                    break
            
            if not block_to_update:
                return {"error": f"Block not found: {block_id}"}
            
            # Replace block content in original content
            # This is a simplified approach - in practice, you'd want more precise replacement
            old_block_content = block_to_update["content"]
            updated_content = content.replace(old_block_content, new_content, 1)
            
            note_path.write_text(updated_content, encoding="utf-8")
            
            return {
                "success": True,
                "note_name": note_path.stem,
                "block_id": block_id,
                "old_content": old_block_content,
                "new_content": new_content
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def embed_note(target_note: str, source_note: str, heading: Optional[str] = None, append: bool = True) -> dict:
        """
        Embed one note into another.
        
        Args:
            target_note: Note to embed into (without .md extension)
            source_note: Note to embed (without .md extension)
            heading: Optional heading to embed specific section
            append: If True, append embed; if False, prepend
        
        Returns:
            Dictionary with embed status
        """
        try:
            vault_path = get_vault_path()
            target_path = safe_path(vault_path, f"{target_note}.md")
            
            if not target_path.exists():
                target_path = safe_path(vault_path, target_note)
            
            if not target_path.exists():
                return {"error": f"Target note not found: {target_note}"}
            
            # Check if source note exists
            source_path = safe_path(vault_path, f"{source_note}.md")
            if not source_path.exists():
                source_path = safe_path(vault_path, source_note)
            
            if not source_path.exists():
                return {"error": f"Source note not found: {source_note}"}
            
            content = target_path.read_text(encoding="utf-8")
            
            # Create embed string
            if heading:
                embed_str = f"![[{source_note}#{heading}]]"
            else:
                embed_str = f"![[{source_note}]]"
            
            # Check if already embedded
            if embed_str in content:
                return {
                    "success": True,
                    "message": f"Note {source_note} is already embedded in {target_note}",
                    "embed": embed_str
                }
            
            # Add embed
            if append:
                updated_content = content.rstrip() + f"\n{embed_str}"
            else:
                updated_content = f"{embed_str}\n{content}"
            
            target_path.write_text(updated_content, encoding="utf-8")
            
            return {
                "success": True,
                "target_note": target_note,
                "source_note": source_note,
                "embed": embed_str
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def get_note_embeds(note_name: str) -> dict:
        """
        Get all notes embedded in a note.
        
        Args:
            note_name: Name of the note (without .md extension) or relative path
        
        Returns:
            Dictionary with embedded notes
        """
        try:
            vault_path = get_vault_path()
            note_path = safe_path(vault_path, f"{note_name}.md")
            
            if not note_path.exists():
                note_path = safe_path(vault_path, note_name)
            
            if not note_path.exists():
                return {"error": f"Note not found: {note_name}"}
            
            content = note_path.read_text(encoding="utf-8")
            embeds = extract_embeds(content)
            
            return {
                "note_name": note_path.stem,
                "total_embeds": len(embeds),
                "embeds": embeds
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def remove_embed(target_note: str, source_note: str, heading: Optional[str] = None) -> dict:
        """
        Remove an embedded note from another note.
        
        Args:
            target_note: Note containing the embed (without .md extension)
            source_note: Embedded note to remove (without .md extension)
            heading: Optional heading if removing specific section embed
        
        Returns:
            Dictionary with removal status
        """
        try:
            vault_path = get_vault_path()
            target_path = safe_path(vault_path, f"{target_note}.md")
            
            if not target_path.exists():
                target_path = safe_path(vault_path, target_note)
            
            if not target_path.exists():
                return {"error": f"Target note not found: {target_note}"}
            
            content = target_path.read_text(encoding="utf-8")
            
            # Create embed string to remove
            if heading:
                embed_str = f"![[{source_note}#{heading}]]"
            else:
                embed_str = f"![[{source_note}]]"
            
            if embed_str not in content:
                return {
                    "success": True,
                    "message": f"Embed not found: {embed_str}",
                    "target_note": target_note
                }
            
            # Remove embed (handle newlines)
            updated_content = content.replace(embed_str, "").strip()
            # Clean up multiple newlines
            updated_content = re.sub(r'\n{3,}', '\n\n', updated_content)
            
            target_path.write_text(updated_content, encoding="utf-8")
            
            return {
                "success": True,
                "target_note": target_note,
                "source_note": source_note,
                "removed_embed": embed_str
            }
        except Exception as e:
            return {"error": str(e)}

File: obsidian-mcp-ts/src/obsidian/embeds.ts
import fs from "fs";
import path from "path";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { getVaultPath } from "./vault.js";
import { resolveNotePath, relPath, stem } from "./utils.js";

interface Block {
  id: string;
  type: string;
  content: string;
  line_number: number;
  level?: number;
  language?: string;
}

function extractBlocks(content: string): Block[] {
  const blocks: Block[] = [];
  const lines = content.split("\n");
  let current: Block | null = null;
  let blockId = 0;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (/^#{1,6}\s/.test(line)) {
      if (current) blocks.push(current);
      const level = line.length - line.replace(/^#+/, "").length;
      current = { id: `block_${blockId}`, type: "heading", level, content: line, line_number: i + 1 };
      blockId++;
    } else if (line.trim().startsWith("```")) {
      if (current?.type === "code") {
        current.content += "\n" + line;
        blocks.push(current);
        current = null;
      } else {
        if (current) blocks.push(current);
        const language = line.trim().slice(3) || "";
        current = { id: `block_${blockId}`, type: "code", language, content: line, line_number: i + 1 };
        blockId++;
      }
    } else if (line.trim().startsWith(">")) {
      if (current?.type === "blockquote") {
        current.content += "\n" + line;
      } else {
        if (current) blocks.push(current);
        current = { id: `block_${blockId}`, type: "blockquote", content: line, line_number: i + 1 };
        blockId++;
      }
    } else {
      if (current) {
        current.content += "\n" + line;
      } else {
        current = { id: `block_${blockId}`, type: "paragraph", content: line, line_number: i + 1 };
        blockId++;
      }
    }
  }

  if (current) blocks.push(current);
  return blocks;
}

function extractEmbeds(content: string): { note: string; heading: string | null; embed_string: string }[] {
  const embeds: { note: string; heading: string | null; embed_string: string }[] = [];
  const pattern = /!\[\[([^\]#]+)(?:#([^\]]+))?\]\]/g;
  for (const m of content.matchAll(pattern)) {
    const note = m[1].trim();
    const heading = m[2]?.trim() || null;
    embeds.push({ note, heading, embed_string: `![[${note}${heading ? "#" + heading : ""}]]` });
  }
  return embeds;
}

export function registerEmbedTools(server: McpServer) {
  server.tool(
    "get_note_blocks",
    "Extract block structure from a note",
    { note_name: z.string().describe("Name of the note (without .md extension) or relative path") },
    async ({ note_name }) => {
      try {
        const vaultPath = getVaultPath();
        const notePath = resolveNotePath(vaultPath, note_name);
        if (!notePath) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Note not found: ${note_name}` }) }] };

        const content = fs.readFileSync(notePath, "utf-8");
        const blocks = extractBlocks(content);

        const blockCounts: Record<string, number> = {};
        for (const b of blocks) blockCounts[b.type] = (blockCounts[b.type] || 0) + 1;

        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ note_name: stem(notePath), total_blocks: blocks.length, block_counts: blockCounts, blocks }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "update_block",
    "Update a specific block in a note",
    {
      note_name: z.string().describe("Name of the note (without .md extension) or relative path"),
      block_id: z.string().describe("ID of the block to update (from get_note_blocks)"),
      new_content: z.string().describe("New content for the block"),
    },
    async ({ note_name, block_id, new_content }) => {
      try {
        const vaultPath = getVaultPath();
        const notePath = resolveNotePath(vaultPath, note_name);
        if (!notePath) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Note not found: ${note_name}` }) }] };

        const content = fs.readFileSync(notePath, "utf-8");
        const blocks = extractBlocks(content);
        const block = blocks.find((b) => b.id === block_id);

        if (!block) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Block not found: ${block_id}` }) }] };

        const updated = content.replace(block.content, new_content);
        fs.writeFileSync(notePath, updated, "utf-8");

        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ success: true, note_name: stem(notePath), block_id, old_content: block.content, new_content }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "embed_note",
    "Embed one note into another",
    {
      target_note: z.string().describe("Note to embed into (without .md extension)"),
      source_note: z.string().describe("Note to embed (without .md extension)"),
      heading: z.string().optional().describe("Optional heading to embed specific section"),
      append: z.boolean().default(true).describe("If true, append embed; if false, prepend"),
    },
    async ({ target_note, source_note, heading, append }) => {
      try {
        const vaultPath = getVaultPath();
        const targetPath = resolveNotePath(vaultPath, target_note);
        if (!targetPath) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Target note not found: ${target_note}` }) }] };

        const sourcePath = resolveNotePath(vaultPath, source_note);
        if (!sourcePath) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Source note not found: ${source_note}` }) }] };

        const content = fs.readFileSync(targetPath, "utf-8");
        const embedStr = heading ? `![[${source_note}#${heading}]]` : `![[${source_note}]]`;

        if (content.includes(embedStr)) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ success: true, message: `Note ${source_note} is already embedded in ${target_note}`, embed: embedStr }) }] };
        }

        const updated = append ? content.trimEnd() + `\n${embedStr}` : `${embedStr}\n${content}`;
        fs.writeFileSync(targetPath, updated, "utf-8");

        return { content: [{ type: "text" as const, text: JSON.stringify({ success: true, target_note, source_note, embed: embedStr }) }] };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "get_note_embeds",
    "Get all notes embedded in a note",
    { note_name: z.string().describe("Name of the note (without .md extension) or relative path") },
    async ({ note_name }) => {
      try {
        const vaultPath = getVaultPath();
        const notePath = resolveNotePath(vaultPath, note_name);
        if (!notePath) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Note not found: ${note_name}` }) }] };

        const content = fs.readFileSync(notePath, "utf-8");
        const embeds = extractEmbeds(content);

        return { content: [{ type: "text" as const, text: JSON.stringify({ note_name: stem(notePath), total_embeds: embeds.length, embeds }) }] };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "remove_embed",
    "Remove an embedded note from another note",
    {
      target_note: z.string().describe("Note containing the embed (without .md extension)"),
      source_note: z.string().describe("Embedded note to remove (without .md extension)"),
      heading: z.string().optional().describe("Optional heading if removing specific section embed"),
    },
    async ({ target_note, source_note, heading }) => {
      try {
        const vaultPath = getVaultPath();
        const targetPath = resolveNotePath(vaultPath, target_note);
        if (!targetPath) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Target note not found: ${target_note}` }) }] };

        const content = fs.readFileSync(targetPath, "utf-8");
        const embedStr = heading ? `![[${source_note}#${heading}]]` : `![[${source_note}]]`;

        if (!content.includes(embedStr)) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ success: true, message: `Embed not found: ${embedStr}`, target_note }) }] };
        }

        let updated = content.replace(embedStr, "").trim();
        updated = updated.replace(/\n{3,}/g, "\n\n");
        fs.writeFileSync(targetPath, updated, "utf-8");

        return { content: [{ type: "text" as const, text: JSON.stringify({ success: true, target_note, source_note, removed_embed: embedStr }) }] };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );
}
